### ISSUE TITLE
Compilation errors on prerelease PowerPC Mac OS X Snow Leopard

### ISSUE THREAD

**Issue by barracuda156** [libass/libass#595]
Compilation fails on this:
```
In file included from ass_utils.c:30:
ass.h:421: error: wrong number of arguments specified for ‘deprecated’ attribute
ass.h:520: error: wrong number of arguments specified for ‘deprecated’ attribute
In file included from ass.c:35:
ass.h:421: error: wrong number of arguments specified for ‘deprecated’ attribute
ass.h:520: error: wrong number of arguments specified for ‘deprecated’ attribute
libtool: compile:  /usr/bin/gcc-4.2 -std=gnu99 -DHAVE_CONFIG_H -I. -I.. -I/opt/local/include -std=gnu99 -Wall -Wextra -Wno-sign-compare -Wno-unused-parameter -Werror-implicit-function-declaration -Wstrict-prototypes -Wpointer-arith -Wredundant-decls -Wno-missing-field-initializers -D_GNU_SOURCE -pipe -Os -arch ppc -I/opt/local/include/freetype2 -I/opt/local/include/libpng16 -I/opt/local/include/fribidi -I/opt/local/include/harfbuzz -I/opt/local/include/glib-2.0 -I/opt/local/lib/glib-2.0/include -I/opt/local/include/freetype2 -I/opt/local/include/libpng16 -I/opt/local/include/ossp -I/opt/local/include/freetype2 -I/opt/local/include/libpng16 -MT ass_string.lo -MD -MP -MF .deps/ass_string.Tpo -c ass_string.c -o ass_string.o >/dev/null 2>&1
make[2]: *** [ass_utils.lo] Error 1
make[2]: *** Waiting for unfinished jobs....
libtool: compile:  /usr/bin/gcc-4.2 -std=gnu99 -DHAVE_CONFIG_H -I. -I.. -I/opt/local/include -std=gnu99 -Wall -Wextra -Wno-sign-compare -Wno-unused-parameter -Werror-implicit-function-declaration -Wstrict-prototypes -Wpointer-arith -Wredundant-decls -Wno-missing-field-initializers -D_GNU_SOURCE -pipe -Os -arch ppc -I/opt/local/include/freetype2 -I/opt/local/include/libpng16 -I/opt/local/include/fribidi -I/opt/local/include/harfbuzz -I/opt/local/include/glib-2.0 -I/opt/local/lib/glib-2.0/include -I/opt/local/include/freetype2 -I/opt/local/include/libpng16 -I/opt/local/include/ossp -I/opt/local/include/freetype2 -I/opt/local/include/libpng16 -MT ass_strtod.lo -MD -MP -MF .deps/ass_strtod.Tpo -c ass_strtod.c -o ass_strtod.o >/dev/null 2>&1
mv -f .deps/ass_string.Tpo .deps/ass_string.Plo
make[2]: *** [ass.lo] Error 1
```

The older version builds:
```
36-47% port -v installed libass
The following ports are currently installed:
  libass @0.13.7_0 (active) requested_variants='' platform='darwin 10' archs='ppc' date='2022-02-14T20:14:23+0800'
```

**Comment by barracuda156** [comment]
If using gcc7, build fails a bit further on this:
```
ass_coretext.c: In function 'is_postscript_font_format':
ass_coretext.c:69:25: error: 'kCTFontFormatOpenTypePostScript' undeclared (first use in this function); did you mean 'kCTFontFormatPostScript'?
         ret = format == kCTFontFormatOpenTypePostScript ||
                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                         kCTFontFormatPostScript
ass_coretext.c:69:25: note: each undeclared identifier is reported only once for each function it appears in
make[2]: *** [ass_coretext.lo] Error 1
make[2]: *** Waiting for unfinished jobs....
mv -f .deps/ass_render.Tpo .deps/ass_render.Plo
mv -f .deps/ass_blur.Tpo .deps/ass_blur.Plo
```

**Comment by barracuda156** [comment]
In fact it builds once `kCTFontFormatOpenTypePostScript` is removed, so that `ass_coretext.c` has:

```
static void destroy_font(void *priv)
{
    CTFontDescriptorRef fontd = priv;
    CFRelease(fontd);
}

static bool is_postscript_font_format(CFNumberRef cfformat)
{
    bool ret = false;
    int format;
    if (CFNumberGetValue(cfformat, kCFNumberIntType, &format)) {
        ret = format == kCTFontFormatPostScript;
    }
    return ret;
}

static bool check_postscript(void *priv)
{
    CTFontDescriptorRef fontd = priv;
    CFNumberRef cfformat =
        CTFontDescriptorCopyAttribute(fontd, kCTFontFormatAttribute);
    if (!cfformat)
        return false;
    bool ret = is_postscript_font_format(cfformat);
    CFRelease(cfformat);
    return ret;
}
```

Not sure if it's a good solution. 

Earlier version had the following:
```
static void destroy_font(void *priv)
{
    CTFontDescriptorRef fontd = priv;
    SAFE_CFRelease(fontd);
}

static bool check_postscript(void *priv)
{
    CTFontDescriptorRef fontd = priv;
    CFNumberRef cfformat =
        CTFontDescriptorCopyAttribute(fontd, kCTFontFormatAttribute);
    int format;

    if (!CFNumberGetValue(cfformat, kCFNumberIntType, &format))
        return false;

    SAFE_CFRelease(cfformat);

    return format == kCTFontFormatOpenTypePostScript ||
           format == kCTFontFormatPostScript;
}
```

Can anyone either confirm if omitting `kCTFontFormatOpenTypePostScript` is okay or something better can be done?

**Comment by TheOneric** [comment]
> >  wrong number of arguments specified for ‘deprecated’ attribute
>
> Which compiler is that?

The log from the OP suggests it’s GCC 4.2 and indeed `git log --grep=deprecated --grep=attribute --all-match releases/gcc-4.2.0..releases/gcc-6.3.0` suggests the parameter was added in [GCC's commit 9b86d6](https://gcc.gnu.org/git/?p=gcc.git;a=commit;h=9b86d6bb25587db93a322bf5778e9892aaa8b776) first included in a release with GCC 4.5. *(GCC 6.3 being the oldest we regularly test in our CI)*

@barracuda156 Can you test with the following applied to confirm compilation otherwise works fine with GCC 4.2?
```diff
diff --git a/libass/ass.h b/libass/ass.h
index dd3833c..df79ef4 100644
--- a/libass/ass.h
+++ b/libass/ass.h
@@ -31,7 +31,11 @@ extern "C" {
 #endif
 
 #if (defined(__GNUC__) && ((__GNUC__ > 3) || (__GNUC__ == 3 && __GNUC_MINOR__ >= 1))) || defined(__clang__)
-    #define ASS_DEPRECATED(msg) __attribute__((deprecated(msg)))
+	#if __GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 5) || defined(__clang__)
+	    #define ASS_DEPRECATED(msg) __attribute__((deprecated(msg)))
+	#else
+		#define ASS_DEPRECATED(msg) __attribute__((deprecated))
+	#endif
     #if __GNUC__ > 5 || defined(__clang__)
         #define ASS_DEPRECATED_ENUM(msg) __attribute__((deprecated(msg)))
     #else
```

**Comment by barracuda156** [comment]
> The `ports` quote suggests you’re running Snow Leopard, which should have `kCTFontFormatOpenTypePostScript`; but AFAICT Snow Leopard didn’t officially support PowerPC. For what it’s worth, is it right to assume this is one of those unofficial macOS distributions/tweaks that adds extended hardware support to newer versions?

@astiob Yes, I was too quick to assume `libass` fails “for PowerPC”, in fact it only fails on Snow Leopard for PowerPC. The logs above are from 10.6 PPC (10A190), it is the latest publicly available developer build working on PowerPC.
I think it also failed on 10.6.8 Rosetta, but I need to reconfirm that when I am back to Intel machine with it.

On 10.5.8 and even 10.4.11 the current version of `libass` builds “as-is”.

10A190 indeed lacks some things which are present in 10.6.8 (which answers your question). So at times it requires additional fixes.

**Comment by barracuda156** [comment]
> As for:
> 
> > wrong number of arguments specified for ‘deprecated’ attribute
> 
> Which compiler is that? The LLVM-backed GCC, or an actual early Clang? We’d be happy to add a check to avoid this error, but we’d need to know how to detect it. :(

@astiob Both `clang` and `llvm-gcc42` are broken for PowerPC (regardless of OS). Initial errors are with `gcc-4.2` (Apple default compiler on 10.6). `kCTFontFormatOpenTypePostScript` error is with `gcc7` – but it is specific to 10.6.

**Comment by barracuda156** [comment]
> @barracuda156 Can you test with the following applied to confirm compilation otherwise works fine with GCC 4.2? _(The patch assumes clang always supported a parameter for `deprecated` which may or may not be true)_

@TheOneric Sure, I will try and update you.

**Comment by astiob** [comment]
> The log from the OP suggests it’s GCC 4.2 and indeed […]

Right, thanks. I’d misread our code and incorrectly assumed that we already handled real GCC’s versioning and this could only happen on Apple’s builds.

---

@barracuda156 I’ve obtained a copy of the 10A190 build and see its version of the Core Text headers. But so I can avoid setting up a PowerPC virtual machine, could you please try the following to test how it actually behaves:

1. Apply this small patch on top of your changes to dump the exact font format values:

    ```diff
     static bool is_postscript_font_format(CFNumberRef cfformat)
     {
         bool ret = false;
         int format;
         if (CFNumberGetValue(cfformat, kCFNumberIntType, &format)) {
    +        printf("font format: %d\n", format);
             ret = format == kCTFontFormatPostScript;
         }
         return ret;
     }
    ```

2. Install Manrope-Regular.ttf from https://fonts.google.com/specimen/Manrope (just a random free TrueType-flavour OpenType font I had around)

3. Play/render this ASS file and post the standard output here:

    ```
    [Script Info]
    
    [V4+ Styles]
    Style: Default,Monaco,18,&H00FFFFFF,,,,,,,,100,100,,,,,,,,,,
    
    [Events]
    Dialogue: 0,0:00:00.00,9:00:00.00,,,,,,,Monaco\N{\fnHiraMinProN-W3}HiraMinProN-W3\N{\fnHelveticaLTMM}HelveticaLTMM{\fnManrope Regular}Manrope Regular
    ```

    (Put this in a `.ass` file and use a video player like mpv or libass’s own `test/test` utility, which is built after `./configure --enable-test`.)

**Comment by astiob** [comment]
Sorry; I put the wrong patch in. Correct patch:

```diff
--- a/libass/ass_coretext.c
+++ b/libass/ass_coretext.c
@@ -147,6 +147,13 @@ static bool get_font_info_ct(CTFontDescriptorRef fontd,
     if (!family_name)
         return false;
 
+    CFNumberRef cfformat =
+        CTFontDescriptorCopyAttribute(fontd, kCTFontFormatAttribute);
+    int format;
+    CFNumberGetValue(cfformat, kCFNumberIntType, &format);
+    printf("font format: %d\n", format);
+    CFRelease(cfformat);
+
     return true;
 }
 
```

**Comment by TheOneric** [comment]
> I’d misread our code and incorrectly assumed that we already handled real GCC’s versioning

Well, we did already *try* to handle all “real GCC” versions. It just didn't occur to me when I originally checked when the attribute was introduced that it might not not always have supported an argument.
 1)  Added attribute: [GCC e23bd21 (January 2002)](https://gcc.gnu.org/git/?p=gcc.git;a=commit;h=e23bd2185fae41772d012231e005b789b3a9e6dc) (first tagged release: 3.1.0 which is **not** the first 3.x.x release)
 2) Accept argument: [GCC 9b86d6b (May 2009)](https://gcc.gnu.org/git/?p=gcc.git;a=commit;h=9b86d6bb25587db93a322bf5778e9892aaa8b776) (first tagged release: 4.5.0)
 3) Extended to enums: [GCC fd5c817 (May 2015)](https://gcc.gnu.org/git/?p=gcc.git;a=commit;h=fd5c817a2457050649c2aadd8657091b2f6bbdaf) (first tagged release: 6.1.0 which **is** the first 6.x.x release)

**Comment by barracuda156** [comment]
@TheOneric I have added the patch above, removing my patch, and tried to build with `gcc-4.2`: no, it still fails (on 10.6 PPC):
```
ass_coretext.c: In function ‘is_postscript_font_format’:
ass_coretext.c:69: error: ‘kCTFontFormatOpenTypePostScript’ undeclared (first use in this function)
ass_coretext.c:69: error: (Each undeclared identifier is reported only once
ass_coretext.c:69: error: for each function it appears in.)
make[2]: *** [ass_coretext.lo] Error 1
make[2]: *** Waiting for unfinished jobs....
mv -f .deps/ass_blur.Tpo .deps/ass_blur.Plo
make[2]: Leaving directory `/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_libass/libass/work/libass-libass-98f5f3e/libass'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_libass/libass/work/libass-libass-98f5f3e'
make: *** [all] Error 2
make: Leaving directory `/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_libass/libass/work/libass-libass-98f5f3e'
Command failed:  cd "/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_libass/libass/work/libass-0.15.2" && /usr/bin/make -j4 -w all 
Exit code: 2
Error: Failed to build libass: command execution failed
Error: See /opt/local/var/macports/logs/_opt_PPCSnowLeopardPorts_multimedia_libass/libass/main.log for details.
Error: Follow https://guide.macports.org/#project.tickets if you believe there
is a bug.
Error: Processing of port libass failed
```

To be precise, this was the only patch used:
```
--- libass/ass.h.orig	2021-09-12 05:34:52.000000000 +0800
+++ libass/ass.h	2022-02-19 23:14:23.000000000 +0800
@@ -31,7 +31,11 @@
 #endif
 
 #if (defined(__GNUC__) && ((__GNUC__ > 3) || (__GNUC__ == 3 && __GNUC_MINOR__ >= 1))) || defined(__clang__)
-    #define ASS_DEPRECATED(msg) __attribute__((deprecated(msg)))
+    #if __GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 5) || defined(__clang__)
+        #define ASS_DEPRECATED(msg) __attribute__((deprecated(msg)))
+    #else
+        #define ASS_DEPRECATED(msg) __attribute__((deprecated))
+    #endif
     #if __GNUC__ > 5 || defined(__clang__)
         #define ASS_DEPRECATED_ENUM(msg) __attribute__((deprecated(msg)))
     #else
```

**Comment by barracuda156** [comment]
@astiob With your patch on top of mine and using `gcc7` it builds.
```
--- libass/ass_coretext.c.orig	2021-09-12 05:34:52.000000000 +0800
+++ libass/ass_coretext.c	2022-02-19 23:24:30.000000000 +0800
@@ -66,8 +66,7 @@
     bool ret = false;
     int format;
     if (CFNumberGetValue(cfformat, kCFNumberIntType, &format)) {
-        ret = format == kCTFontFormatOpenTypePostScript ||
-              format == kCTFontFormatPostScript;
+        ret = format == kCTFontFormatPostScript;
     }
     return ret;
 }
@@ -146,6 +145,12 @@
     info->extended_family = family_name;
     if (!family_name)
         return false;
+    CFNumberRef cfformat =
+        CTFontDescriptorCopyAttribute(fontd, kCTFontFormatAttribute);
+    int format;
+    CFNumberGetValue(cfformat, kCFNumberIntType, &format);
+    printf("font format: %d\n", format);
+    CFRelease(cfformat);
 
     return true;
 }
```

I will do testing in a day or two. I have built it now with `--enable-test`, but need to figure out how to use it. Spent past 15 hrs with fixing `ghc` for PowerPC.

**Comment by barracuda156** [comment]
@TheOneric Yes, with both patches together it does build with `gcc-4.2`, if that is what you wanted to know.

**Comment by astiob** [comment]
> I have built it now with `--enable-test`, but need to figure out how to use it.

I believe you just pass in a target output PNG image file path, an input subtitle file and a timestamp. For this test, this should do:

```sh
test/test /dev/null test.ass 0
```

Take your time, and thanks in advance.

**Comment by TheOneric** [comment]
@barracuda156: did you manage to run the test?

**Comment by barracuda156** [comment]
@TheOneric Sorry, this issue got postponed back then, and I forgot about it. I only have 10.6.8 Rosetta at the moment (away from PPC hardware), and on Rosetta `libass` build fails badly – I will need to fix that first (this part should not be hard though).

I will make an update here in a couple of days re status on Rosetta; with 10.6 PPC I will be able to test it a bit later.

**Comment by barracuda156** [comment]
@TheOneric A quick update: build on Rosetta is fixed in this: https://github.com/macports/macports-ports/commit/574c9c4e30d2ff12a120fb05a9da11e0d997cf7b
However I need to get back to my PPC machines to test it natively on 10.5.8 and 10.6 PPC. Please allow me until the end of month: it is on my to-do list once I am back to Taiwan.

**Comment by barracuda156** [comment]
@astiob @TheOneric Apologize, it took ages to get back. I am getting this:
```
36-43% /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_libass/libass/work/libass-libass-9971ce6/test/test -o /Users/svacchanda/Dev/libass/test.ass test
test.c: Available font providers (4): 'None', 'Autodetect', 'CoreText', 'Fontconfig'.
libass: libass API version: 0x1600000
libass: libass source: tarball: 0.16.0
libass: Raster: FreeType 2.12.1
libass: Shaper: FriBidi 1.0.12 (SIMPLE) HarfBuzz-ng 5.2.0 (COMPLEX)
libass: Initialized
libass: Using font provider coretext
libass: File size: 250
libass: [0x41060b0] Style: Default,Monaco,18,&H00FFFFFF,,,,,,,,100,100,,,,,,,,,,
libass: No event format found, using fallback
libass: [0x41060b0]: Warning: no style named '' found, using 'Default'
libass: Added subtitle file: '/Users/svacchanda/Dev/libass/test.ass' (2 styles, 1 events)
libass: Neither PlayResX nor PlayResY defined. Assuming 384x288
font format: 2
libass: fontselect: (Monaco, 400, 0) -> /System/Library/Fonts/Monaco.ttf, -1, Monaco
font format: 1
libass: fontselect: (HiraMinProN-W3, 400, 0) -> /System/Library/Fonts/ヒラギノ明朝 ProN W3.otf, -1, HiraMinProN-W3
font format: 3
font format: 3
libass: fontselect: (HelveticaLTMM, 400, 0) -> /System/Library/Fonts/HelveLTMM, -1, HelveticaLTMM
font format: 1
libass: fontselect: (Manrope Regular, 400, 0) -> /Users/svacchanda/Library/Fonts/manrope/manrope-regular.otf, -1, Manrope-Regular
4 images blended
```
Is this something sensible? :)

**Comment by astiob** [comment]
Thanks! Not sure why `font format: 3` is printed twice, but it seems clear enough to conclude the following:

* Interestingly enough, Type 1 fonts are actually accessible via this build of Core Text. Now that I think about it, we probably get Type 1 fonts from GDI on Windows and IIRC we do get them from Fontconfig too (I should note this on [the wiki page](https://github.com/libass/libass/wiki/Fonts-across-platforms) for future reference), but we don’t have any special handling for them. What actually happens when we get one?

* OpenType fonts with quadratic (TrueType) and cubic (PostScript) outlines cannot be distinguished via Core Text itself. Above I said this was critical, but now that I think about it, we immediately parse each Core Text–supplied font via FreeType anyway in order to fetch the font’s names. We should be able to check the outline type via FreeType at the same time.

**Comment by barracuda156** [comment]
@astiob Thank you for responding! I can and will provide any further input that may be of use, if you kindly instruct on what to test in particular. My understanding of fonts is from pre-press background, which is not what is required here :)

Testing is of no issue (two PowerMacs with everything set for building), and I have incentive to pursue this issue to be fixed with upstream, since Macports are not accepting 10.6 PPC-specific fixes (95%+ of problems of 10.6 PPC are common with 10.5.8 and/or Rosetta, but few are unique).

**Comment by barracuda156** [comment]
@astiob Can I make the PR with the fixes above? (Of course, making those conditional.)
Or can we improve upon those?

**Comment by astiob** [comment]
Assuming you mean effectively just removing the `kCTFontFormatOpenTypePostScript` line, no, it just makes it _compile_ but at runtime the font selection code doesn’t do what it’s supposed to do. The proper fix is to do this:

> we immediately parse each Core Text–supplied font via FreeType anyway in order to fetch the font’s names. We should be able to check the outline type via FreeType at the same time.

In fact, this would even let us support Mac OS X 10.5 by additionally using [this incantation](https://github.com/harfbuzz/harfbuzz/blob/72c6962012fab7a4bfe345687d10f8d08c672151/src/hb-coretext.cc#L204-L214) to get the font’s file path.

I’ve submitted https://github.com/libass/libass/pull/664 for this. If you happen to have a 10.5 machine available for testing (no matter whether Intel or PowerPC), I’d appreciate it if you could try it there, too. The same test ASS file as above should do nicely, and it’s enough to just verify that the output image looks the same.

As a no-longer-relevant reminder, on any Mac you can also use Fontconfig instead of the Core Text–based backend: install Fontconfig, and when building libass on the 10A190, use `./configure --disable-coretext`.

**Comment by barracuda156** [comment]
@astiob Yes, I got 10.5, will try it there soon and update. Thank you very much!

**Comment by barracuda156** [comment]
@astiob I have built `libass` now from your 10.5 tree on 10.5.8 as `ppc32`. Build is fine (no extra patches used). Running test sample gives this:
```
36-205:~ svacchanda$ /opt/local/var/macports/build/_opt_PPCLeopardPorts_multimedia_libass/libass/work/astiob-libass-199fce1/test/test -o /Users/svacchanda/Dev/libass/test.ass test
test.c: Available font providers (4): 'None', 'Autodetect', 'CoreText', 'Fontconfig'.
libass: libass API version: 0x1600010
libass: libass source: tarball: 0.16.0
libass: Raster: FreeType 2.12.1
libass: Shaper: FriBidi 1.0.12 (SIMPLE) HarfBuzz-ng 4.3.0 (COMPLEX)
libass: Initialized
libass: Using font provider coretext
libass: File size: 250
libass: [0x705ce0] Style: Default,Monaco,18,&H00FFFFFF,,,,,,,,100,100,,,,,,,,,,
libass: No event format found, using fallback
libass: [0x705ce0]: Warning: no style named '' found, using 'Default'
libass: Added subtitle file: '/Users/svacchanda/Dev/libass/test.ass' (2 styles, 1 events)
libass: Neither PlayResX nor PlayResY defined. Assuming 384x288
libass: fontselect: (Monaco, 400, 0) -> /System/Library/Fonts/Monaco.dfont, -1, Monaco
libass: fontselect: (HiraMinProN-W3, 400, 0) -> /System/Library/Fonts/ヒラギノ明朝 ProN W3.otf, -1, HiraMinProN-W3
libass: fontselect: (HelveticaLTMM, 400, 0) -> /System/Library/Fonts/HelveLTMM, -1, HelveticaLTMM
libass: fontselect: Using default font family: (Manrope Regular, 400, 0) -> /System/Library/Fonts/Helvetica.dfont, -1, Helvetica
libass: fontselect: (Manrope Regular, 400, 0) -> /System/Library/Fonts/Helvetica.dfont, -1, Helvetica
4 images blended
```
So apparently it does not see Manrope. I placed it in the same location: `/Users/svacchanda/Library/Fonts/manrope`.

**Comment by astiob** [comment]
> So apparently it does not see Manrope. I placed it in the same location: `/Users/svacchanda/Library/Fonts/manrope`.

I suspect you must open the font and click Install, rather than just put the file in this directory.

**Comment by barracuda156** [comment]
> > So apparently it does not see Manrope. I placed it in the same location: `/Users/svacchanda/Library/Fonts/manrope`.
> 
> I suspect you must open the font and click Install, rather than just put the file in this directory.

Normally placing a font in a relevant folder works on macOS, I have been doing that for fonts which I wanted to be always available (I never used install for that, AFAIR). On 10.6 it was immediately available, on 10.5.8 I restarted the OS to make sure it will be available after initial failure.
But I will double-check in another app to make sure the OS sees the font.

**Comment by astiob** [comment]
Thanks for double-checking. (I’ve learnt something new.)

Does libass find the font if you change `\fnManrope Regular` to `\fnManrope`?

Meanwhile, I’m seeing weird behaviour on this font with DirectWrite, too… 🤔 GDI properly finds it by “Manrope Regular” and reports it as the font’s full name, but DirectWrite tells libass its full name is just “Manrope”. Wut?

**Comment by barracuda156** [comment]
@astiob With Manrope Bold in Leopard (I edited `test_ass` file accordingly):
```
36-31:~ svacchanda$ /opt/local/var/macports/build/_opt_PPCLeopardPorts_multimedia_libass/libass/work/astiob-libass-199fce1/test/test -o /Users/svacchanda/Desktop/test.ass test
test.c: Available font providers (4): 'None', 'Autodetect', 'CoreText', 'Fontconfig'.
libass: libass API version: 0x1600010
libass: libass source: tarball: 0.16.0
libass: Raster: FreeType 2.12.1
libass: Shaper: FriBidi 1.0.12 (SIMPLE) HarfBuzz-ng 4.3.0 (COMPLEX)
libass: Initialized
libass: Using font provider coretext
libass: File size: 244
libass: [0x705ce0] Style: Default,Monaco,18,&H00FFFFFF,,,,,,,,100,100,,,,,,,,,,
libass: No event format found, using fallback
libass: [0x705ce0]: Warning: no style named '' found, using 'Default'
libass: Added subtitle file: '/Users/svacchanda/Desktop/test.ass' (2 styles, 1 events)
libass: Neither PlayResX nor PlayResY defined. Assuming 384x288
libass: fontselect: (Monaco, 400, 0) -> /System/Library/Fonts/Monaco.dfont, -1, Monaco
libass: fontselect: (HiraMinProN-W3, 400, 0) -> /System/Library/Fonts/ヒラギノ明朝 ProN W3.otf, -1, HiraMinProN-W3
libass: fontselect: (HelveticaLTMM, 400, 0) -> /System/Library/Fonts/HelveLTMM, -1, HelveticaLTMM
libass: fontselect: (Manrope Bold, 400, 0) -> /Users/svacchanda/Library/Fonts/manrope/manrope-bold.otf, -1, Manrope-Bold
4 images blended
```

**Comment by barracuda156** [comment]
@astiob And with just Manrope (also on 10.5.8):
```
test.c: Available font providers (4): 'None', 'Autodetect', 'CoreText', 'Fontconfig'.
libass: libass API version: 0x1600010
libass: libass source: tarball: 0.16.0
libass: Raster: FreeType 2.12.1
libass: Shaper: FriBidi 1.0.12 (SIMPLE) HarfBuzz-ng 4.3.0 (COMPLEX)
libass: Initialized
libass: Using font provider coretext
libass: File size: 234
libass: [0x705cd0] Style: Default,Monaco,18,&H00FFFFFF,,,,,,,,100,100,,,,,,,,,,
libass: No event format found, using fallback
libass: [0x705cd0]: Warning: no style named '' found, using 'Default'
libass: Added subtitle file: '/Users/svacchanda/Desktop/test.ass' (2 styles, 1 events)
libass: Neither PlayResX nor PlayResY defined. Assuming 384x288
libass: fontselect: (Monaco, 400, 0) -> /System/Library/Fonts/Monaco.dfont, -1, Monaco
libass: fontselect: (HiraMinProN-W3, 400, 0) -> /System/Library/Fonts/ヒラギノ明朝 ProN W3.otf, -1, HiraMinProN-W3
libass: fontselect: (HelveticaLTMM, 400, 0) -> /System/Library/Fonts/HelveLTMM, -1, HelveticaLTMM
libass: fontselect: (Manrope, 400, 0) -> /Users/svacchanda/Library/Fonts/manrope/manrope-regular.otf, -1, Manrope-Regular
```

**Comment by astiob** [comment]
> Meanwhile, I’m seeing weird behaviour on this font with DirectWrite, too… 🤔 GDI properly finds it by “Manrope Regular” and reports it as the font’s full name, but DirectWrite proceeds to say its full name is just “Manrope”. Wut?

OK, on DirectWrite, this behaviour is apparently caused by this font’s [`STAT` table](https://learn.microsoft.com/en-us/typography/opentype/spec/stat). Here’s a copy with the STAT table removed: [Manrope-Regular-woSTAT.zip](https://github.com/libass/libass/files/10474935/Manrope-Regular-woSTAT.zip); and DirectWrite reports its full name as “Manrope Regular”. But I doubt Leopard would’ve known anything about this table: it seems to be a pretty recent table, and it’s absent from Apple’s documentation even now.

**Comment by barracuda156** [comment]
@astiob The build from 927565a0365339d8dde2c644a8cb98ff497ff263 is all good on 10.6 PPC.
Output from test file:
```
36-25% /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_libass/libass/work/libass-927565a0365339d8dde2c644a8cb98ff497ff263/test/test -o /Users/svacchanda/Dev/libass/test.ass test
test.c: Available font providers (4): 'None', 'Autodetect', 'CoreText', 'Fontconfig'.
libass: libass API version: 0x1701000
libass: libass source: tarball: 0.17.1
libass: Raster: FreeType 2.12.1
libass: Shaper: FriBidi 1.0.13 (SIMPLE) HarfBuzz-ng 8.1.1 (COMPLEX)
libass: Initialized
libass: Using font provider coretext
libass: File size: 250
libass: [0x4105ff0] Style: Default,Monaco,18,&H00FFFFFF,,,,,,,,100,100,,,,,,,,,,
libass: No event format found, using fallback
libass: [0x4105ff0]: Warning: no style named '' found, using 'Default'
libass: Added subtitle file: '/Users/svacchanda/Dev/libass/test.ass' (2 styles, 1 events)
libass: Neither PlayResX nor PlayResY defined. Assuming 384x288
libass: fontselect: (Monaco, 400, 0) -> /System/Library/Fonts/Monaco.ttf, -1, Monaco
libass: fontselect: (HiraMinProN-W3, 400, 0) -> /System/Library/Fonts/ヒラギノ明朝 ProN W3.otf, -1, HiraMinProN-W3
libass: fontselect: (HelveticaLTMM, 400, 0) -> /System/Library/Fonts/HelveLTMM, -1, HelveticaLTMM
libass: fontselect: Using default font family: (Manrope Regular, 400, 0) -> /System/Library/Fonts/Helvetica.ttc, -1, Helvetica
libass: fontselect: (Manrope Regular, 400, 0) -> /System/Library/Fonts/Helvetica.ttc, -1, Helvetica
4 images blended
```

So here it does not want to see Manrope. FontBook sees it normally and validation returns no errors with the font.

**Comment by barracuda156** [comment]
Just in case, rebuilt from 46255cedfb92b1d339d9b8c90537cac82d31e7be and repeated, identical outcome.

**Comment by barracuda156** [comment]
@astiob Sorry to bother again, could you please take a look? I get the failure now on 10.6 PPC with gcc13:
```
libass/ass_coretext.c: In function 'is_postscript_font_format':
libass/ass_coretext.c:70:25: error: 'kCTFontFormatOpenTypePostScript' undeclared (first use in this function); did you mean 'kCTFontFormatPostScript'?
   70 |         ret = format == kCTFontFormatOpenTypePostScript ||
      |                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                         kCTFontFormatPostScript
libass/ass_coretext.c:70:25: note: each undeclared identifier is reported only once for each function it appears in
libass/ass_coretext.c: In function 'get_font_file':
libass/ass_coretext.c:111:14: error: implicit declaration of function 'CHECK_AVAILABLE' [-Werror=implicit-function-declaration]
  111 |     else if (CHECK_AVAILABLE(kCTFontURLAttribute, macOS 10.6, *)) {
      |              ^~~~~~~~~~~~~~~
libass/ass_coretext.c:111:51: error: 'macOS' undeclared (first use in this function)
  111 |     else if (CHECK_AVAILABLE(kCTFontURLAttribute, macOS 10.6, *)) {
      |                                                   ^~~~~
libass/ass_coretext.c:111:56: error: expected ')' before numeric constant
  111 |     else if (CHECK_AVAILABLE(kCTFontURLAttribute, macOS 10.6, *)) {
      |                             ~                          ^~~~~
      |                                                        )
cc1: some warnings being treated as errors
make[1]: *** [libass/ass_coretext.lo] Error 1
```

**Comment by barracuda156** [comment]
P. S. I think I just miss some other fix from the master – I tried to apply the patch 927565a0365339d8dde2c644a8cb98ff497ff263 to 0.17.1.
Maybe release 0.17.2?

**Comment by barracuda156** [comment]
Ah, this is what I need, I guess 1630dbeeef3b541e86af0c324a53a05a4f5a4b22

**Comment by barracuda156** [comment]
@astiob Still fails:
```
libass/ass_coretext.c: In function 'get_font_file':
libass/ass_coretext.c:88:14: error: implicit declaration of function 'CHECK_AVAILABLE' [-Werror=implicit-function-declaration]
   88 |     else if (CHECK_AVAILABLE(kCTFontURLAttribute, macOS 10.6, *)) {
      |              ^~~~~~~~~~~~~~~
libass/ass_coretext.c:88:51: error: 'macOS' undeclared (first use in this function)
   88 |     else if (CHECK_AVAILABLE(kCTFontURLAttribute, macOS 10.6, *)) {
      |                                                   ^~~~~
libass/ass_coretext.c:88:51: note: each undeclared identifier is reported only once for each function it appears in
libass/ass_coretext.c:88:56: error: expected ')' before numeric constant
   88 |     else if (CHECK_AVAILABLE(kCTFontURLAttribute, macOS 10.6, *)) {
      |                             ~                          ^~~~~
      |                                                        )
cc1: some warnings being treated as errors
make[1]: *** [libass/ass_coretext.lo] Error 1
make[1]: *** Waiting for unfinished jobs....
```

**Comment by astiob** [comment]
@barracuda156 Just build `master`, please. If you need a tarball because Git is unavailable, GitHub provides them: https://github.com/libass/libass/archive/refs/heads/master.tar.gz

**Comment by astiob** [comment]
Ah, and if you’re using Homebrew, IIRC `brew install --HEAD libass` should work if you have `head "https://github.com/libass/libass" in `libass.rb`.

**Comment by astiob** [comment]
If you do absolutely need individual patches, then you need the following commits (from oldest to newest), as seen in https://github.com/libass/libass/pull/664:

1. 1630dbeeef3b541e86af0c324a53a05a4f5a4b22
2. possibly 4eff3a78483f7c956d1109d17fc3e070d06b2820
3. 847af2a1ce67cc2ac0db1252f41ee99ff724b05b
4. 5c15c883a4783641f7e71a6a1f440209965eb64f

**Comment by astiob** [comment]
Although 1630dbeeef3b541e86af0c324a53a05a4f5a4b22 _alone_ should also suffice on your 10.6 in my understanding.

**Comment by barracuda156** [comment]
@astiob Sorry for a delay. `libass` builds fine from the master on 10.6 PowerPC without any patches from my side, at least with the modern GCC. So we are all good here.

The only technical issue I have is how to port the fix to existing released version, since Macports does not build it from master, and introducing `libass-devel` port is perhaps unjustified otherwise. I will try your suggestions here.
